### DISCUSSION TITLE
Why is the Racket cask saying it's being depreciated upon installing?

### DISCUSSION THREAD

**Discussion by andrewcrook** [Homebrew/discussions#3510] [Casks] [answered]
The racket cask seems to be still maintained.

When you install it you get a warning saying that it is being depreciated and to use minimum-racket formula
However, the cask is the quickest and easiest way to install the whole racket and drracket environment.
In my point of view this cask should stay and the depreciation warning removed.

may be it should be renamed to drracket or full-racket

The minimum-racket formula should also remain for those who want to use on the commandline, other IDES and cherry pick features using its own package manager raco.

**Comment by carlocab** [comment]
I'm not sure why you're seeing a deprecation warning, but I don't think it's coming from Homebrew.

What is the output of
```
brew info --cask racket
brew cat --cask racket
```
?

**Comment by gromgit** [comment]
What you're seeing is almost certainly because you tried to install the `racket` _formula_ rather than the cask:
```
$ brew install racket
Warning: Use minimal-racket instead of deprecated racket
Warning: Use minimal-racket instead of deprecated racket
Warning: Treating racket as a formula. For the cask, use homebrew/cask/racket
[...]
```
As you can see, `brew` is quite explicit about what happened, so you probably missed that last warning line.

To install the `racket` _cask_: `brew install --cask racket`.

**Reply by andrewcrook** [reply]
> What you're seeing is almost certainly because you tried to install the `racket` _formula_ rather than the cask:
> 

@gromgit 

Nope I am installing the cask and I have used homebrew for years.

```
>> brew install racket --cask                                                                                      
Warning: Use minimal-racket instead of deprecated racket
==> Downloading https://mirror.racket-lang.org/installers/8.6/racket-8.6-aarch64-macosx-cs.dmg
Already downloaded: /Users/XXXXXX/Library/Caches/Homebrew/downloads/c85e232ada01f00b012a4ec3328432946321f525e83693f931c62fafa8c84907--racket-8.6-aarch64-macosx-cs.dmg
==> Installing Cask racket
...
...
```

```
brew install --cask  racket                                                                                
Warning: Use minimal-racket instead of deprecated racket
==> Downloading https://mirror.racket-lang.org/installers/8.6/racket-8.6-aarch64-macosx-cs.dmg
Already downloaded: /Users/XXXXXX/Library/Caches/Homebrew/downloads/c85e232ada01f00b012a4ec3328432946321f525e83693f931c62fafa8c84907--racket-8.6-aarch64-macosx-cs.dmg
==> Installing Cask racket
.....
....
```

**Reply by andrewcrook** [reply]
@gromgit  

Interesting I found this method is different but installs the same diskimage, no warning 

If I use `brew install homebrew/cask/racket` no warning but surely this is the same as `brew install --cask racket`?


```
brew install homebrew/cask/racket                                                                             
Running `brew update --auto-update`...
==> Auto-updated Homebrew!
==> Updated Homebrew from 402e267a0 to de6e848db.
Updated 5 taps (homebrew/cask-versions, homebrew/command-not-found, homebrew/core, homebrew/cask and blacktop/tap).
==> New Casks
gittyup                                                                          imhex

You have 2 outdated formulae installed.
You can upgrade them with brew upgrade
or list them with brew outdated.

==> Downloading https://mirror.racket-lang.org/installers/8.6/racket-8.6-aarch64-macosx-cs.dmg
Already downloaded: /Users/XXXXXX/Library/Caches/Homebrew/downloads/c85e232ada01f00b012a4ec3328432946321f525e83693f931c62fafa8c84907--racket-8.6-aarch64-macosx-cs.dmg
==> Installing Cask racket
....
....
```

**Comment by cho-m** [comment]
Guessing that brew is still loading the `homebrew/core` tap when passing `--cask`.

Can reproduce and changing warning to exception generates following stacktrace:
```
/usr/local/Homebrew/Library/Homebrew/formulary.rb:461:in `formula_name_path'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:429:in `initialize'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:685:in `new'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:685:in `loader_for'
/usr/local/Homebrew/Library/Homebrew/formulary.rb:552:in `factory'
/usr/local/Homebrew/Library/Homebrew/cli/parser.rb:626:in `block in formulae'
/usr/local/Homebrew/Library/Homebrew/cli/parser.rb:622:in `map'
/usr/local/Homebrew/Library/Homebrew/cli/parser.rb:622:in `formulae'
/usr/local/Homebrew/Library/Homebrew/cli/parser.rb:307:in `parse'
/usr/local/Homebrew/Library/Homebrew/cmd/install.rb:145:in `install'
```

Maybe from https://github.com/Homebrew/brew/blob/3.6.0/Library/Homebrew/cli/parser.rb#L302-L307 which tries to parse all input as formulae. Not familiar enough with code to see where we can restrict behavior when `--cask`/`--casks` is used.
```rb
        if @formula_options
          remaining, non_options = parse_remaining(argv, ignore_invalid_options: true)

          argv = [*remaining, "--", *non_options]

          formulae(argv).each do |f|
```

**Comment by andrewcrook** [comment]
I have tried different DMGs and source as well as minimum-racket, racket cask on brew and asdf-racket via asdf 

minimum-racket you have to make sure installs the MacOS precompiled dynlibs

Tbh, It is just better to install racket cask in the first place which is drracket, racket, command line tools, common teaching packages and is the Chez Scheme version.

Don't know why I get the depreciation warning its certainly not the case.

**Comment by johnnaaron** [comment]
I also get the "Use minimal-racket instead of deprecated racket" when running "brew install --cask racket" right now

**Comment by andrewcrook** [comment]
Hi @gromgit @johnaaron-git @cho-m 

A bit of an update...

I believe the cause is that `homebrew-cask` is indirectly using `homebrew-core/formula_renames.json`

[line:](https://github.com/Homebrew/homebrew-core/blob/d9baeae8fe1420738683e8b0532cd5cca59565dd/formula_renames.json#L130) `  "racket": "minimal-racket”,`

There used to be a racket.rb formula which was replaced with minimal-racket.rb

The racket.rb cask is been compared to this list so is picked up as the old formula 
where homebrew-cask should possibly have its own cask_renames.json list in the future?

Still it is a very rare issue perhaps `"racket": "minimal-racket”,` can be removed from `homebrew-core/formula_renames.json` now some time has passed?

**Reply by andrewcrook** [reply]
I dont think you get the message when the DMG file is still in the download cache

**Reply by muzimuzhi** [reply]
> I believe the cause is that `homebrew-cask` is indirectly using `homebrew-core/formula_renames.json`
> 
> [line:](https://github.com/Homebrew/homebrew-core/blob/d9baeae8fe1420738683e8b0532cd5cca59565dd/formula_renames.json#L130) ` "racket": "minimal-racket”,`

Sounds promising. That line was added when formula `racket` was renamed to `minimal-racket` in Homebrew/homebrew-core#3735, but then in Homebrew/homebrew-cask#23553 a `racket` cask was (re)introduced thus caused a name crash.

**Reply by muzimuzhi** [reply]
Testing locally, deleting that line in `formula_renames.json` in my local `homebrew-cask/core` clone then `brew install --cask racket` installs the `racket` cask successfully.

Is it recommended to provide a PR or request change of logic in ruby scripts?

I'm slightly tend to the latter one. I recalled that in the past `brew cask install racket` did worked. I can't remember if I successfully installed `racket` through `brew install --cask racket` before.

---
To restate the problem, currently there're
- a `minimal-racket` formula once called `racket`, hence a lin `"racket": "minimal-racket"` in `homebrew-core/formula_renames.json` and
- a `racket` cask

Then run `brew install --cask racket`,
- Expected: `--cask` has higher priority than `homebrew-core/formula_renames.json`, thus the cask `racket` is installed.
- Actual: `--cask` has <s>higher</s> lower priority than `homebrew-core/formula_renames.json`, thus the formula `minimal-cask` is installed, after the message `Warning: Use minimal-racket instead of deprecated racket`.

It seems to me this should be reported to https://github.com/Homebrew/brew.

**Reply by andrewcrook** [reply]
@muzimuzhi   Many thanks I too verified this but haven’t had time to raise a PR. Just remove it sounds the best solution now sufficient time has passed after renaming to minimal-racket. Although casks could have their own list as a new feature. I think the former would do in this scenario.

**Reply by andrewcrook** [reply]
@muzimuzhi Oh yes, could you please post the PR link here for reference . Thanks.

**Reply by muzimuzhi** [reply]
After posting the previous comment I then realized that it might be a bug which could be resolved without modifying `formula_renames.json`. You may need to visit GitHub webpage to see the edited version of my comment.

Basically, if the program is coded well, then `racket` cask and a renaming entry in `formula_renames.json` could live together:
- `brew install minimal-racket` installs `minimal-racket` formula;
- `brew install racket` installs `minimal-racket` formula as well, after throwing `Warning: Use minimal-racket instead of deprecated racket`;
- `brew install --cask racket` installs `racket` cask.

The only problem is that currently entry in `formula_renames.json` has higher priority than `--cask`.

**Reply by muzimuzhi** [reply]
> Just remove it sounds the best solution now sufficient time has passed after renaming to minimal-racket.

Sorry I overlooked your first comment. I'm still a little hesitant about providing a PR to remove `"racket": "minimal-racket",` brutally. But yes, that would be the quickest fix, if not the most elegant.

**Reply by jryans** [reply]
Yes, I would think it's reasonable to just remove the renaming... It was [first added](https://github.com/Homebrew/homebrew-core/commit/25a696d4c1e1bf9ec1509dd5f43f6f5df5715967) in 2016, so it's been around for quite a while. At this point, I feel the rename is just causing confusion (since the cask is the best choice for most).

**Comment by muzimuzhi** [comment, marked as answer]
Maybe I'm not the only one who was misled by the warning
```
Warning: Use minimal-racket instead of deprecated racket
```

I double checked that `brew install --cask racket` does **have cask `racket` installed**, not the aliased formula `minimal-racket`. This can be verified in several ways:
- In the stdout of `brew install --cask racket`, except for the warning shown above
    ```
    ==> Downloading https://mirror.racket-lang.org/installers/8.7/racket-8.7-x86_64-macosx-cs.
    ######################################################################## 100.0%
    ==> Installing Cask racket
    ==> Moving App Suite 'Racket v8.7' to '/Applications/Racket v8.7'
    ==> Linking Binary 'drracket' to '/usr/local/bin/drracket'
    ...
    ==> Linking Manpage 'drracket.1' to '/usr/local/share/man/man1/drracket.1'
    ...
    🍺  racket was successfully installed!
    ```
- Run `brew info minimal-racket` to check formula `minimal-racket` is not installed.
- Run `brew info --cask racket` or rerun `brew install --cask racket` to check the cask `racket` is installed.
- Noting cask `racket` downloads a dmg and formula downloads the source code `racket-minimal-8.7-src.tgz` and in homebrew's cache directory (`brew --cache`) no `racket-minimal` related cache is found.

So the remaining problem is, homebrew should not check formula renaming list when `--cask` is given. I'll report it to [`Homebrew/brew`](https://github.com/Homebrew/brew).

**Reply by muzimuzhi** [reply]
see https://github.com/Homebrew/brew/issues/14438

**Reply by muzimuzhi** [reply]
> see [Homebrew/brew#14438](https://github.com/Homebrew/brew/issues/14438)

Update: It has been resolved by Homebrew/brew#14450. Now (after `brew update`) `brew install --cask --dry-run racket` won't throw that warning.

BTW, @andrewcrook is it possible to change the comment chosen as answer?

**Reply by andrewcrook** [reply]
@muzimuzhi 

>  BTW, @andrewcrook is it possible to change the comment chosen as answer?

Done
